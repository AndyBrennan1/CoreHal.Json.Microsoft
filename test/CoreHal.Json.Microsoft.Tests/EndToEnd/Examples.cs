using System;
using System.Collections.Generic;
using System.Linq;
using Xunit;
using System.Text.Json;
using ApprovalTests.Reporters;
using ApprovalTests;
using CoreHal.Graph;
using CoreHal.PropertyNaming;
using CoreHal.Json.Tests.EndToEnd.SimpleEcommerceDomain;

namespace CoreHal.Json.Tests.EndToEnd
{
    [UseReporter(typeof(VisualStudioReporter))]
    [Collection("End to End Examples")]
    public class Examples
    {
        public HalGraph IHalGraph { get; private set; }

        [Fact]
        public void Example1()
        {
            var model = new Order
            {
                CustomerName = "A B Systems",
                OrderDate = new DateTime(2020, 6, 23),
                OrderId = Guid.Parse("AA957846-2448-42DC-A1CB-90E9E399F57C"),
                OrderNumber = "ORDER123",
                OrderValue = 6.5M
            };

            var customer = new Customer
            {
                CustomerId = Guid.Parse("ACE2A06E-5FE5-44DE-BDFC-AAFC9C507DCC"),
                CompanyName = "A-B-Systems",
                PrimaryContact = "Andy Brennan",
                PrimaryPhoneNumber = "+44787642628292",
                Sector = new Sector
                {
                    SectorId = Guid.Parse("030A522A-EF75-44F3-AF61-62FA8FA655A9"),
                    Name = "NuGet Making"
                }
            };

            var orderLines = new List<OrderLine>
            {
                new OrderLine {
                    OrderLineId = Guid.Parse("999cc766-5370-4d66-9aa8-6ae02787762c"),
                    ProductName = "Apple",
                    SKU = "APP123",
                    LineValue = 1.0M,
                    Quantity = 2
                },
                new OrderLine {
                    OrderLineId = Guid.Parse("950c1b91-38a3-413c-9d00-b886ea435943"),
                    ProductName = "Orange",
                    SKU = "ORA555",
                    LineValue = 1.5M,
                    Quantity = 2
                },
                new OrderLine {
                    OrderLineId = Guid.Parse("e258854b-9a6b-4ff0-86c2-a8fe2589d276"),
                    ProductName = "Pear",
                    SKU = "PER88",
                    LineValue = 3.0M,
                    Quantity = 6
                }
            };

            IHalGraph customerGraph =
                new HalGraph(customer)
                .AddLink("self", new Link("/api/customers/A-B-Systems"));

            IHalGraph trySingleOrderLine =
                new HalGraph(orderLines[0])
                .AddLink("order", new Link("/api/orders/ORDER123"))
                .AddLink("parent", new Link("/api/orders/ORDER123/order-lines"))
                .AddLink("self", new Link($"/api/orders/ORDER123/order-lines/{orderLines[0].OrderLineId}"));

            IEnumerable<IHalGraph> orderlineGraphs = orderLines.Select(
                orderLine => new HalGraph(orderLine)
                .AddLink("order", new Link("/api/orders/ORDER123"))
                .AddLink("parent", new Link("/api/orders/ORDER123/order-lines"))
                .AddLink("self", new Link($"/api/orders/ORDER123/order-lines/{orderLine.OrderLineId}")));

            IHalGraph orderGraph =
                new HalGraph(model)
                .AddCurie(new Curie("ord", "/api/orders/{order-number}"))
                .AddLink("self", new Link("/api/orders/ORDER123"))
                .AddEmbeddedItem("customer", customerGraph)
                .AddEmbeddedItem("singular-order-line", trySingleOrderLine)
                .AddEmbeddedItemCollection("order-lines", orderlineGraphs);

            IPropertyNameConvention propertyNameConvention = new CamelDashNameConvention();

            var serializerOptions = new JsonSerializerOptions
            {
                Converters =
                {
                    new HalGraphJsonConvertor(propertyNameConvention),
                    new HalGraphCollectionJsonConvertor(propertyNameConvention),
                    new LinkJsonConvertor(propertyNameConvention),
                    new CurieLinkJsonConvertor(propertyNameConvention)
                },
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
                WriteIndented = true
            };

            var responseJson = JsonSerializer.Serialize(orderGraph, typeof(HalGraph), serializerOptions);

            Approvals.Verify(responseJson);
        }
    }
}
