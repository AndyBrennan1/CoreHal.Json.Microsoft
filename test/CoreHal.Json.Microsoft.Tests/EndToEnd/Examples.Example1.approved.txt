{
  "_links": {
    "curies": [
      {
        "name": "ord",
        "href": "/api/orders/{order-number}",
        "templated": true
      }
    ],
    "self": {
      "href": "/api/orders/ORDER123"
    }
  },
  "order-Id": "aa957846-2448-42dc-a1cb-90e9e399f57c",
  "customer-Name": "A B Systems",
  "order-Date": "2020-06-23T00:00:00",
  "order-Value": 6.5,
  "order-Number": "ORDER123",
  "_embedded": {
    "customer": {
      "_links": {
        "self": {
          "href": "/api/customers/A-B-Systems"
        }
      },
      "customer-Id": "ace2a06e-5fe5-44de-bdfc-aafc9c507dcc",
      "company-Name": "A-B-Systems",
      "primary-Phone-Number": "\u002B44787642628292",
      "primary-Contact": "Andy Brennan",
      "sector": {
        "SectorId": "030a522a-ef75-44f3-af61-62fa8fa655a9",
        "Name": "NuGet Making"
      }
    },
    "singular-order-line": {
      "_links": {
        "order": {
          "href": "/api/orders/ORDER123"
        },
        "parent": {
          "href": "/api/orders/ORDER123/order-lines"
        },
        "self": {
          "href": "/api/orders/ORDER123/order-lines/999cc766-5370-4d66-9aa8-6ae02787762c"
        }
      },
      "order-Line-Id": "999cc766-5370-4d66-9aa8-6ae02787762c",
      "product-Name": "Apple",
      "sku": "APP123",
      "quantity": 2,
      "line-Value": 1.0
    },
    "order-lines": [
      {
        "_links": {
          "order": {
            "href": "/api/orders/ORDER123"
          },
          "parent": {
            "href": "/api/orders/ORDER123/order-lines"
          },
          "self": {
            "href": "/api/orders/ORDER123/order-lines/999cc766-5370-4d66-9aa8-6ae02787762c"
          }
        },
        "order-Line-Id": "999cc766-5370-4d66-9aa8-6ae02787762c",
        "product-Name": "Apple",
        "sku": "APP123",
        "quantity": 2,
        "line-Value": 1.0
      },
      {
        "_links": {
          "order": {
            "href": "/api/orders/ORDER123"
          },
          "parent": {
            "href": "/api/orders/ORDER123/order-lines"
          },
          "self": {
            "href": "/api/orders/ORDER123/order-lines/950c1b91-38a3-413c-9d00-b886ea435943"
          }
        },
        "order-Line-Id": "950c1b91-38a3-413c-9d00-b886ea435943",
        "product-Name": "Orange",
        "sku": "ORA555",
        "quantity": 2,
        "line-Value": 1.5
      },
      {
        "_links": {
          "order": {
            "href": "/api/orders/ORDER123"
          },
          "parent": {
            "href": "/api/orders/ORDER123/order-lines"
          },
          "self": {
            "href": "/api/orders/ORDER123/order-lines/e258854b-9a6b-4ff0-86c2-a8fe2589d276"
          }
        },
        "order-Line-Id": "e258854b-9a6b-4ff0-86c2-a8fe2589d276",
        "product-Name": "Pear",
        "sku": "PER88",
        "quantity": 6,
        "line-Value": 3.0
      }
    ]
  }
}